src/main/java/hudson/remoting/AsyncFutureImpl.java

```diff
@@ -78,13 +79,28 @@ public synchronized V get() throws InterruptedException, ExecutionException {
         return value;
     }
 
+    @CheckForNull
     public synchronized V get(long timeout, TimeUnit unit) throws InterruptedException, ExecutionException, TimeoutException {
-        if(!completed)
-            wait(unit.toMillis(timeout));
+        long timeoutMs = unit.toMillis(timeout);
+        // It is not very accurate, but all(proof?) production uses of this class pass seconds.
+        // Test cases pass 1 millisecond and other such fun, but wait(long) is not time-accurate anyway
+        long endWaitTime = System.currentTimeMillis() + timeoutMs;
+        while (!completed) {
+            long currentTime = System.currentTimeMillis();
+            if (currentTime > endWaitTime) {
+                break;
+            }
+            wait(endWaitTime - currentTime);
+        }
+
         if(!completed)
             throw new TimeoutException();
         if(cancelled)
             throw new CancellationException();
+
+        //TODO: we may be calling a complex get() implementation in the override, which may hang the code
+        // The only missing behavior is "problem!=null" branch, but probably we cannot just replace the code without
+        // an override issue risk.
```

Comment:
Then perhaps move the get outside the synchronised

Reply:
I will need to investigate the implementations first. `get()` outside synchronized may help for sure, but I would like to find real examples before touching this code.

Reply:
Ack